presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalSumAggregation.java

```diff
@@ -124,48 +124,54 @@ private static InternalAggregationFunction generateAggregation(Type inputType, T
+    public static void inputShortDecimal(Type type, LongDecimalWithOverflowState state, Block block, int position)
     {
-        accumulateValueInState(BigInteger.valueOf(type.getLong(block, position)), state);
+        accumulateValueInState(unscaledDecimal(type.getLong(block, position)), state);
     }
 
-    public static void inputLongDecimal(Type type, BigIntegerState state, Block block, int position)
+    public static void inputLongDecimal(Type type, LongDecimalWithOverflowState state, Block block, int position)
     {
-        accumulateValueInState(decodeUnscaledValue(type.getSlice(block, position)), state);
+        accumulateValueInState(type.getSlice(block, position), state);
     }
 
-    private static void accumulateValueInState(BigInteger value, BigIntegerState state)
+    private static void accumulateValueInState(Slice unscaledDecimal, LongDecimalWithOverflowState state)
     {
         initializeIfNeeded(state);
-        state.setBigInteger(state.getBigInteger().add(value));
+        Slice sum = state.getLongDecimal();
+        long overflow = UnscaledDecimal128Arithmetic.addReturnOverflow(sum, unscaledDecimal, sum);
+        state.setOverflow(state.getOverflow() + overflow);
     }
 
-    private static void initializeIfNeeded(BigIntegerState state)
+    private static void initializeIfNeeded(LongDecimalWithOverflowState state)
     {
-        if (state.getBigInteger() == null) {
-            state.setBigInteger(BigInteger.valueOf(0));
+        if (state.getLongDecimal() == null) {
+            state.setLongDecimal(UnscaledDecimal128Arithmetic.unscaledDecimal());
         }
     }
 
-    public static void combine(BigIntegerState state, BigIntegerState otherState)
+    public static void combine(LongDecimalWithOverflowState state, LongDecimalWithOverflowState otherState)
     {
-        if (state.getBigInteger() == null) {
-            state.setBigInteger(otherState.getBigInteger());
+        state.setOverflow(state.getOverflow() + otherState.getOverflow());
+
+        if (state.getLongDecimal() == null) {
+            state.setLongDecimal(otherState.getLongDecimal());
         }
         else {
-            state.setBigInteger(state.getBigInteger().add(otherState.getBigInteger()));
+            accumulateValueInState(otherState.getLongDecimal(), state);
         }
     }
 
-    public static void outputLongDecimal(DecimalType type, BigIntegerState state, BlockBuilder out)
+    public static void outputLongDecimal(DecimalType type, LongDecimalWithOverflowState state, BlockBuilder out)
     {
-        if (state.getBigInteger() == null) {
+        if (state.getLongDecimal() == null) {
             out.appendNull();
         }
         else {
-            BigDecimal value = new BigDecimal(state.getBigInteger(), type.getScale());
-            checkOverflow(state.getBigInteger());
-            writeBigDecimal(type, out, value);
+            if (state.getOverflow() != 0) {
```

Comment:
One scenario when that can happen is for aggregating following values:
```
[MAX_VALUE/2, MAX_VALUE/2, MAX_VALUE/2, (-MAX_VALUE/2), (-MAX_VALUE/2), (-MAX_VALUE/2)]
```
Both input and combine can overflow over max value temporarily, but we can salvage situation if accumulated value drops below `MAX_VALUE` later.